### PR TITLE
fix bleed and ignite critical strike proportions to be based on how many applications you can apply during the duration

### LINKED ISSUE #4719: Critical bleed does not take into account stacks
### Check version

- [X] I'm running the latest version of Path of Building and I've verified this by checking the [changelog](https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/master/CHANGELOG.md)

### Check for duplicates

- [X] I've checked for duplicate issues by using the search function of the [issue tracker](https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues)

### What is the expected behaviour/value?

In the calculation for bleed there is a calculation that averages between the proportion of hits caused by critical damage vs teh proportion caused by non critical damage.
Somehow the stack potential value should come into play here, as only the strongest bleed will come into account, so this should not be an average between crit damage and non crit damage, but rather a function of stack potential, Where with a stack potential of 1 you get the current equation, and with a stack potential of infinite you have all the damage from critical hits.

Failing anything else a configuration option to allow all ailments to be calculated for criticals would also be useful.

### What is the actual behaviour/value?

Currently there is a flat split of damage between the crit and the non crit values, strongly underestimating bleed damage with high duration/low crit chance

### How to reproduce the issue

Selevt puncture as main skill.
Spec into somewhat strong critical build and perfect agony
With 25% crit, crit multi nodes/perfect agony seem much weaker than expected

### Character build code

```shell
2.19.2
```

### Screenshots

[bleedcalc](https://user-images.githubusercontent.com/43895880/183657657-6ee7683b-420c-45b1-a5dc-466f7efc9636.PNG)
)

### PR DESCRIPTION
**Pull request by oljomo**
Increase critical strike chance used in damage calculation for stacking DoTs (ignite and bleed) based on how far over the stack limit you are.

Fixes #4719 .

### Steps taken to verify a working solution:
-
-
-

### CODE REVIEW

File: src/Modules/CalcOffence.lua [hunk 1]
```diff
@@ -3097,7 +3097,9 @@ function calcs.offence(env, actor, activeSkill)
 				end
 			end
 			local basePercent = skillData.bleedBasePercent or data.misc.BleedPercentBase
-			local baseVal = calcAilmentDamage("Bleed", sourceHitDmg, sourceCritDmg) * basePercent / 100 * output.RuthlessBlowBleedEffect * output.FistOfWarAilmentEffect * globalOutput.AilmentWarcryEffect
+			-- overstacking bleed stacks increases the chance a critical ignite is present
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:3100]:
```suggestion
			-- overstacking bleed stacks increases the chance a critical bleed is present
```

File: src/Modules/CalcOffence.lua [hunk 2]
```diff
@@ -3280,7 +3283,8 @@ function calcs.offence(env, actor, activeSkill)
 					s_format("Ailment mode: %s ^8(can be changed in the Configuration tab)", igniteMode == "CRIT" and "Crits Only" or "Average Damage")
 				}
 			end
-			local baseVal = calcAilmentDamage("Poison", sourceHitDmg, sourceCritDmg) * data.misc.PoisonPercentBase * output.FistOfWarAilmentEffect * globalOutput.AilmentWarcryEffect
+			local ailmentCritChance = output.CritChance
+			local baseVal = calcAilmentDamage("Poison", ailmentCritChance, sourceHitDmg, sourceCritDmg) * data.misc.PoisonPercentBase * output.FistOfWarAilmentEffect * globalOutput.AilmentWarcryEffect
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:3287]:
```suggestion
			local baseVal = calcAilmentDamage("Poison", output.CritChance, sourceHitDmg, sourceCritDmg) * data.misc.PoisonPercentBase * output.FistOfWarAilmentEffect * globalOutput.AilmentWarcryEffect
```

File: src/Modules/CalcOffence.lua [hunk 3]
```diff
@@ -3733,7 +3739,8 @@ function calcs.offence(env, actor, activeSkill)
 					s_format("Ailment mode: %s ^8(can be changed in the Configuration tab)", igniteMode == "CRIT" and "Crits Only" or "Average Damage")
 				}
 			end
-			local baseVal = calcAilmentDamage("Freeze", calcAverageSourceDamage("Freeze")) * skillModList:More(cfg, "FreezeAsThoughDealing")
+			local ailmentCritChance = output.CritChance
+			local baseVal = calcAilmentDamage("Freeze", ailmentCritChance, calcAverageSourceDamage("Freeze")) * skillModList:More(cfg, "FreezeAsThoughDealing")
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:3743]:
```suggestion
			local baseVal = calcAilmentDamage("Freeze", output.CritChance, calcAverageSourceDamage("Freeze")) * skillModList:More(cfg, "FreezeAsThoughDealing")
```

File: src/Modules/CalcOffence.lua [hunk 4]
```diff
@@ -3751,7 +3758,8 @@ function calcs.offence(env, actor, activeSkill)
 						s_format("Ailment mode: %s ^8(can be changed in the Configuration tab)", igniteMode == "CRIT" and "Crits Only" or "Average Damage")
 					}
 				end
-				local damage = calcAilmentDamage(ailment, calcAverageSourceDamage(ailment)) * skillModList:More(cfg, ailment.."AsThoughDealing")
+				local ailmentCritChance = output.CritChance
+				local damage = calcAilmentDamage(ailment, ailmentCritChance, calcAverageSourceDamage(ailment)) * skillModList:More(cfg, ailment.."AsThoughDealing")
```
Comment (QuickStick123) [inline src/Modules/CalcOffence.lua:3762]:
```suggestion
				local damage = calcAilmentDamage(ailment, output.CritChance, calcAverageSourceDamage(ailment)) * skillModList:More(cfg, ailment.."AsThoughDealing")
```

**Comment by QuickStick123** [comment]
local chanceFromCrit = output.BleedChanceOnCrit / 100 * output.CritChance / 100
Needs to use ailmentCritChance for bleed and ignite.

**Review by LocalIdentity — APPROVED** [review]

**Pull request merged**